lib/class-wp-annotation-utils.php

```diff
@@ -0,0 +1,820 @@
+		$allow_statuses = array_merge( $allow_statuses, self::$custom_statuses );
+
+		if ( $allow_actions ) {
+			$allow_statuses = array_merge( $allow_statuses, array( 'unspam', 'untrash' ) );
+		}
+		return is_string( $status ) && in_array( $status, $allow_statuses, true );
+	}
+
+	/**
+	 * Checks an annotation comment status (or status action).
+	 *
+	 * @since [version]
+	 *
+	 * @param  string $status Comment status.
+	 *
+	 * @return bool           True if status is valid.
+	 */
+	public static function is_valid_status_or_action( $status ) {
+		return self::is_valid_status( $status, true );
+	}
+
+	/**
+	 * Validates a W3C annotation client identifier.
+	 *
+	 * @since [version]
+	 *
+	 * @param  string $client The annotation client to check.
+	 *
+	 * @return bool           True if client is valid.
+	 *
+	 * @link https://www.w3.org/TR/annotation-model/#rendering-software
+	 */
+	public static function is_valid_client( $client ) {
+		if ( '' === $client ) {
+			return true; // Empty is OK.
+		}
+
+		if ( ! is_string( $client ) ) {
+			return false;
+		}
+		$raw_client = $client;
+		$client     = preg_replace( '/[^a-z0-9:_\-]/i', '', $client );
+		$client     = substr( trim( $client, ':_-' ), 0, 250 );
+
+		if ( ! $client || $client !== $raw_client ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Validates a W3C annotation selector deeply.
+	 *
+	 * @since [version]
+	 *
+	 * @param  array $selector  Selector to check.
+	 * @param  bool  $recursive For internal use only.
+	 *
+	 * @return bool             True if selector is valid.
+	 *
+	 * @link https://www.w3.org/TR/annotation-model/#selectors
+	 */
+	public static function is_valid_selector( $selector, $recursive = false ) {
+		if ( ! $recursive && array() === $selector ) {
+			return true; // Empty array is OK.
+		}
+
+		if ( ! $selector || ! is_array( $selector ) ) {
+			return false;
+		} elseif ( empty( $selector['type'] ) || ! is_string( $selector['type'] ) ) {
+			return false;
+		} elseif ( 2 > count( array_keys( $selector ) ) ) {
+			return false;
+		}
+
+		/**
+		 * Filters selector types allowed for annotations.
+		 *
+		 * @since [version]
+		 *
+		 * @param array Selector types allowed for annotations.
+		 */
+		$allow_selectors = apply_filters( 'annotation_allow_selectors', self::$allow_selectors );
+		if ( ! in_array( $selector['type'], $allow_selectors, true ) ) {
+			return false;
+		}
+
+		if ( 'RangeSelector' !== $selector['type'] ) {
+			if ( 'SvgSelector' === $selector['type'] ) {
+				$max_selector_size = 131072; // 128kb.
+			} else {
+				$max_selector_size = 16384; // 16kb.
+			}
+
+			/**
+			 * Filters max annotation selector size (in bytes).
+			 *
+			 * @since [version]
+			 *
+			 * @param int    Max annotation selector size (in bytes).
+			 * @param array  An array of all selector details.
+			 */
+			$max_selector_size = apply_filters( 'annotation_max_selector_size', $max_selector_size, $selector );
```

Comment:
No, the [spec](https://www.w3.org/TR/annotation-model/#selectors) doesn't call for a maximum, but it probably should, because selectors can be nested using [refinements](https://www.w3.org/TR/annotation-model/#refinement-of-selection). Some selector types are allowed to contain text, etc. So the goal here is to set an upper limit on size in addition to type-specific validation that occurs below.